src/rpc/RPCClient.ts

```diff
@@ -185,31 +240,98 @@ class RPCClient<M extends ClientManifest> {
     };
   }
 
+  /**
+   * Generic caller for duplex RPC calls.
+   * This returns a `ReadableWritablePair` of the types specified. No validation
+   * is applied to these types so make sure they match the types of the handler
+   * you are calling.
+   * When finished the streams must be ended manually. Failing to do so will
+   * hold the connection open and result in a resource leak until the
+   * call times out.
+   * @param method - Method name of the RPC call
+   * @param ctx - ContextTimed used for timeouts and cancellation.
+   */
   @ready(new rpcErrors.ErrorRPCDestroyed())
   public async duplexStreamCaller<I extends JSONValue, O extends JSONValue>(
     method: string,
+    ctx: Partial<ContextTimed> = {},
   ): Promise<ReadableWritablePair<O, I>> {
+    const abortController = new AbortController();
+    const signal = abortController.signal;
+    // A promise that will reject if there is an abort signal or timeout
+    const abortRaceProm = promise<never>();
+    abortRaceProm.p.catch(() => {});
+    let abortHandler: () => void;
+    if (ctx.signal != null) {
+      // Propagate signal events
+      abortHandler = () => {
+        abortController.abort(ctx.signal?.reason);
+        abortRaceProm.rejectP(ctx.signal?.reason);
+      };
+      if (ctx.signal.aborted) abortHandler();
+      ctx.signal.addEventListener('abort', abortHandler);
+    }
+    const timer =
+      ctx.timer ??
+      new Timer({
+        delay: this.defaultTimeout,
+      });
+    const cleanUp = () => {
+      // Clean up the timer and signal
+      if (ctx.timer == null) timer.cancel(Error('TMP Clean up reason'));
+      signal.removeEventListener('abort', abortHandler);
+    };
+    const timeoutError = new rpcErrors.ErrorRpcTimedOut();
+    void timer.then(
+      () => {
+        abortController.abort(timeoutError);
+        abortRaceProm.rejectP(timeoutError);
+      },
+      () => {},
+    );
+    // Deciding if we want to allow refreshing
+    // We want to refresh timer if none was provided
+    const refreshingTimer: Timer | undefined =
+      ctx.timer == null ? timer : undefined;
     // Providing empty metadata here. we don't support it yet.
     const outputMessageTransformStream =
-      rpcUtils.clientOutputTransformStream<O>({
-        nodeId: IdInternal.fromBuffer<NodeId>(Buffer.alloc(32, 0)), // FIXME
-        host: '',
-        port: 0,
-        command: method,
-      });
-    const inputMessageTransformStream =
-      rpcUtils.clientInputTransformStream<I>(method);
-    const middleware = this.middlewareFactory();
+      rpcUtils.clientOutputTransformStream<O>(
+        {
+          nodeId: IdInternal.fromBuffer<NodeId>(Buffer.alloc(32, 0)), // FIXME
+          host: '',
+          port: 0,
```

Comment:
It's placeholder for some metadata used then creating errors. I haven't found a good way to get the information we want here. It's something I need to review and come back to.

Reply:
Not all errors should require a host and port then?

Reply:
All remote errors should, we'd want to know where the error came from. 

The problem here is, the information is supplied from the client side, but given that the `RPCClient` is agnostic to the transport. We don't have this information within the `RPCClient`, we shouldn't have it. so I don't have a nice way of supplying this information right now. 

Two ways of solving this come to mind.
1. The error revival could be done as a middleware, the metadata can be passed to the middleware and bypass the `RPCClient` internals that way.
2. The metadata can be supplied by the server side, but we'd have a similar problem, So it would have to be part of the middleware as well I think.

Reply:
We can reuse the `RPCStream` to do this:

```ts
interface RPCStream extends ReadableWritablePair {
  cancel(reason?: any) => void;
  meta?: POJO;
}
```

Then this `streamFactory` which produces the `RPCStream` interface, is the only bridge between the transport domain and the RPC domain.

That's the only place to provide any data about the stream.

Reply:
Metadata such as this is provided by the `RPCStream` metadata now. This is done for both the server and client.